Rust: add wrapper for cbindgen (#13093)
mesonbuild/meson: docs/markdown/Rust-module.md

bindgen_clang_arguments = ['--target', 'x86_64-linux-gnu']
```

## cbindgen()

*Since 1.12.0*

```meson

> file_h = rustmod.cbindgen('infile.rs', 'outfile.h', config : 'config.toml')

Reviewer: The recommended file name is `cbindgen.toml` isn't it? `config.toml` OTOH is the convention for the config file that goes in `source_root/.cargo/`

> file_h = rustmod.cbindgen('infile.rs', 'outfile.h', config : 'config.toml')

Reviewer: Why is a source file needed? cbindgen can also take the current source dir as the input.

Author: Because Meson always works on files, because that's more-or-less a requirement to make Ninja work reliably.

Reviewer: The depfile outputted by cbindgen isn't enough for that?

Author: I would consider that to be a significant design change, in the past we've *always* required some kind of file, and custom_targets() that work on `@CURRENT_SOURCE_DIR@` usually end up with some weird corner cases.

I'd want to see some wider consensus from the group before I'd add that.

Reviewer 2: I agree, there's always time to change it later.

```

This function wraps cbindgen to simplify creating C bindings for Rust targets.
It has several advantages over invoking a `custom_target` directly:

>   - it correctly handles depfile, adding them when supported

Reviewer: ```suggestion
  - Correctly handles depfiles, adding them when supported
```

  - Sets the profile to `debug` when Meson is configured with -Ddebug=true

>   - Passes several useful options for use in Meson, such as quite and cpp-compat

Reviewer: ```suggestion
  - Automatically passes some options for use in Meson, such as `--quiet` and `--cpp-compat`
```

  - Handles structured_sources correctly

>   - will automatically detect language based on file extension (c or c++ only),

Reviewer: ```suggestion
  - Will automatically detect language based on file extension (c or c++ only),
```

The majority of the configuration should be handled inside the `config.toml`
file.

It takes the following positional arguments:

>   - `infile`: the root rust file to bind, or a structured_sources instance

Reviewer: ```suggestion
  - `infile`: The root rust file to bind, or a [`structured_sources()`](Reference-manual_functions_structured_sources.md) instance
```

>   - `outfile`: the headerto generate

Reviewer: ```suggestion
  - `outfile`: The header to generate
```

It takes the following keyword arguments

  - `config`: The Path to a configuration toml. May be a File, str, or CustomTarget

>   - `langauge`: The language to bind for. If unset, Meson will choose a language

Reviewer: ```suggestion
  - `language`: The language to bind for. If unset, Meson will choose a language
```

    based on the file extension of the output file. Currently, auto detection only
    works for C and C++. May be one of: `c`, `cpp`, `cython`.
  - `depends`: An array of CustomTargets that this target depends on

>   - `depend_files`: An array of files that this target depends on

Reviewer: Also need kwargs for installation: `install` and `install_dir`. Should also set `install_tag: 'devel'` automatically.
